Split geometric shape handlers into Controller-View

bhennion opened the pull request:
This PR is a proposed first step towards splitting gui/PageView into a view and a controller.
Before this split could be possible, we need to split every tool handler: gui/PageView owns tool handlers which will need to be split into a view, owned by a future view/PageView, and a handler, owned by a future control/PageController.

In this PR, we split the derived classes of BaseStrokeHandler (i.e. the handlers for all geometric shapes -- Rectangle, ellipse, arrow and so on -- but not StrokeHandler). A lot of planning has gone into this PR. I'll try my best to describe the plan, stop me if you object to any parts of it.

- The gui/PageView has a pool of `OverlayView`'s, which should, in time, contain what I call overlays: everything that needs to be painted but is not actually in the page model (e.g. current tools, selections, search results highlighting). For now, this pool contains at most 1 element, corresponding, if there is one, to the active BaseStrokeHandler.
- In order to be able, in a distant future, to have several views of the current document (e.g. to implement #1638 or #2020), each BaseStrokeHandler now has a pool of views, and dispatches repaint requests to the entire pool. Similar things should be implemented for other tools (and other overlays) in the future.
- The OverlayView's inherit from a base class Repaintable, which should, in due time, take over gui/Redrawable. ~~The class Repaintable includes a possible change of coordinates, so that repaint request are always expressed in local coordinates, and transfered up to a parent Repaintable after changing from child coordinates to parent coordinates (think of selections, pages in a layout, or maybe one day, subpages #3684).~~ Edit: This was removed in an iteration of the PR.
- Once this strategy is applied to all other tool handlers (at least StrokeHandler, SplineHandler, SetsquareHandler, Selection, EditSelection, PdfElemSelection, VerticalToolHandler, SearchControl), then we can actually split gui/PageView into a view and a controller. The view view/PageView would inherit from Repaintable, and control/PageController would have a pool of view/PageView's to send repaint/rerender/change requests to.

More specific implementation comments:
- To accomodate the various dispatch needs, a very generic DispatchPool class is implemented. (I'd love feedback on the implementation here)
- ~~The data of the BaseStrokeHandler's is also now protected by a mutex, to address #3579, at least for those tools.~~
    **Edit:** this is actually not necessary, see #3579.
- The class BaseStrokeHandler is renamed to BaseShapeHandler in order to fix its confusing name.

bhennion commented:
> I'll have to take a closer look at the last commit. That is quite big and needs some time to focus on all parts.

Of course! No hurry there.

> Regarding the dispatch pool: Please avoid macros whenever possible. They lead to really complex bugs that are amazingly hard to debug.

Yeah. My first attempt had no macro, creating a listener base class, but I failed to make something satisfactory.
The issue is the following. Consider the following code
```
template<class T> class DispatchPool {
    template <typename... Args>
    void dispatch(Args&&... args) const {
        for (auto* v: this->pool) {
            v->on(std::forward<Args>(args)...);
        }
    }
    std::vector<T*> pool;
};
template<class T> class Listener {
    void addToPool(DispatchPool<T>*);
    void removeFromPool();
    DispatchPool<T>* pool;
};

class MyClass: public Listener<MyClass>{
   void on(const MyEventType& event);
};
```
This works fine, except `Listener<T>::addToPool` has to recast its `this` to a `T*` in order to add it to the pool. 
This is unsafe unless there is a way (say a static_assert) to make sure that any instantiation of `Listener<T>` is actually of type `T`. Is there such a way?
Of course, I could resolve to a dynamic_cast, but I'd like to get rid of the oveheads.

bhennion commented:
I think I found an ok solution for getting rid of the macro DispatchPool.

Febbe commented on src/core/control/tools/InputHandler.cpp:
```diff
@@ -16,47 +16,44 @@
- * @return Current editing stroke
- */
-auto InputHandler::getStroke() -> Stroke* { return stroke; }
+auto InputHandler::getStroke() const -> Stroke* { return stroke.get(); }
```
It looks like, that stroke will never be initialized anymore, how about removing this then?

Febbe commented on src/util/Range.cpp:
```diff
@@ -1,33 +1,47 @@
+        maxX(std::numeric_limits<double>::lowest()),
+        maxY(std::numeric_limits<double>::lowest()) {}
+
+Range::Range(double x, double y): minX(x), minY(y), maxX(x), maxY(y) {}
```
Move to header, since this does not rely on another header file (faster compile time).
```suggestion
```

Febbe commented on src/util/include/util/Range.h:
```diff
@@ -11,11 +11,18 @@
+    Range();  // Empty range
     Range(double x, double y);
-    virtual ~Range();
+    Range(double minX, double minY, double maxX, double maxY);
```
```suggestion
    Range(double minX, double minY, double maxX, double maxY): minX(minX), minY(minY), maxX(maxX), maxY(maxY) {}
```

Febbe commented on src/util/include/util/Range.h:
```diff
@@ -11,11 +11,18 @@
+class Range final {
 public:
+    Range();  // Empty range
     Range(double x, double y);
```
```suggestion
    Range(double x, double y): minX(x), minY(y), maxX(x), maxY(y) {}
```

Febbe commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,83 @@
+
+namespace xoj::util {
+template <class T>
+class DispatchPool {
+public:
+    using listener_type = T;
+    DispatchPool() = default;
+    DispatchPool(DispatchPool&&) = default;
+    DispatchPool(const DispatchPool&) = default;
+    DispatchPool& operator=(DispatchPool&&) = default;
+    DispatchPool& operator=(const DispatchPool&) = default;
+
```
Either make this class final *and* remove all default special member functions, or add a defaulted virtual destructor

Febbe commented on src/core/view/overlays/ShapeToolView.h:
```diff
@@ -0,0 +1,56 @@
+
+public:
+    ShapeToolView(BaseShapeHandler* toolHandler, const Repaintable* parent);
+    virtual ~ShapeToolView() noexcept;
```
Use override instead of virtual. Also, can this class be defined final?
```suggestion
    ~ShapeToolView() override noexcept;
```

Febbe commented on src/core/view/overlays/ShapeToolView.h:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Listener interface
+     */
+    static constexpr struct PAINT_REQUEST_t {
```
Please use snake_case or PascalCase for types. `snake_case` for STL like/mimic structures. `PascalCase` for Xournal internal classes.

Febbe commented on src/core/model/OverlayBase.h:
```diff
@@ -0,0 +1,17 @@
+
+#pragma once
+
+class OverlayBase {
+public:
+    virtual ~OverlayBase() = default;
+};
```
What's the use case of an empty base class, without virtual functions to call?

bhennion replied:
Wait no! There was a use case: see `XojPageView::deleteViewOf(OverlayBase*)`. I just needed a common pointer type for everything that will be an overlay (for now, only InputHandler, but more should come).

Febbe replied:
Then OverlayBase should named ViewBase, and it should require all view functions defined from Repaintable.

bhennion replied:
No, it's not a view. It's a base for some controllers: those controllers which can have one or more views. This includes all the tool controllers (by that I mean things in control/tools), but also SearchControl for instance (to highlight the search results).

It is obviously not clear or we wouldn't be having this discussion, so something needs to be done to clarify the situation. Not sure what yet.

Febbe commented on src/core/gui/PageView.h:
```diff
@@ -39,8 +41,9 @@ class Element;
+class OverlayBase;
 
-class XojPageView: public Redrawable, public PageListener {
+class XojPageView: public Redrawable, public PageListener, public xoj::view::Repaintable {
```
We already have Redrawable, why we need an additional Repaintable.

Febbe replied:
Ok, how about renaming `Redrawable` -> `LegacyRedrawable` and `Repaintable` -> `Redrawable` ? :D

Febbe replied:
At least the name wouldn't be "stupid" and the other name states, that the type is about to replaced some time.

Febbe commented on src/core/gui/PageView.cpp:
```diff
@@ -302,26 +306,43 @@ auto XojPageView::onButtonPressEvent(const PositionInputData& pos) -> bool {
-            this->inputHandler = new CoordinateSystemHandler(this->xournal, this, getPage());
-        } else {
-            this->inputHandler = new StrokeHandler(this->xournal, this, getPage());
+
+        if (this->inputHandler) {
+            this->deleteViewOf(this->inputHandler);
+            delete this->inputHandler;
+            this->inputHandler = nullptr;
+        }
+
+        BaseShapeHandler* shapeHandler = nullptr;
+        switch (h->getDrawingType()) {
+            case DRAWING_TYPE_LINE:
+                shapeHandler = new RulerHandler(this->xournal, this, getPage());
+                break;
+            case DRAWING_TYPE_RECTANGLE:
+                shapeHandler = new RectangleHandler(this->xournal, this, getPage());
+                break;
+            case DRAWING_TYPE_ELLIPSE:
+                shapeHandler = new EllipseHandler(this->xournal, this, getPage());
+                break;
+            case DRAWING_TYPE_ARROW:
+                shapeHandler = new ArrowHandler(this->xournal, this, getPage(), false);
+                break;
+            case DRAWING_TYPE_DOUBLE_ARROW:
+                shapeHandler = new ArrowHandler(this->xournal, this, getPage(), true);
+                break;
+            case DRAWING_TYPE_COORDINATE_SYSTEM:
+                shapeHandler = new CoordinateSystemHandler(this->xournal, this, getPage());
+                break;
+            default:
+                this->inputHandler = new StrokeHandler(this->xournal, this, getPage());
+        }
+        if (shapeHandler) {
+            this->inputHandler = shapeHandler;
+            this->overlayViews.emplace_back(std::make_unique<xoj::view::ShapeToolView>(shapeHandler, this));
```
It really feels bad that we recreate all Handlers, when a button is pressed. This looks like a huge performance loss.
It would be better to either have a std::variant, containing all. Or have them all pre-allocated and only std::exchange them in.

Febbe replied:
It's often safer to recreate, than to reset a structure, but I would just call something like this:

`*handler = Handler{};`
instead of `handler = new Handler{};`
or `new(handler) Handler{};`

Febbe reviewed (changes requested):
I mostly gave language related feedback, since @LittleHuba is the code owner of this part. So he can give you the best feedback on this.

But I have also a question, did you look at the gtk4 drawing-API? They want to replace Cairo with a node based, approach. Therefore, your changes should not increase the porting effort.

bhennion commented on src/util/Range.cpp:
```diff
@@ -1,33 +1,47 @@
+}
 
-auto Range::getY2() const -> double { return this->y2; }
+auto Range::empty() const -> bool { return this->minX > this->maxX || this->minY > this->maxY; }
```
I'm confused. First, is could be `invalid`, maybe, but certainly not `valid`.
Secondly, a Range corresponds to the area of points (x,y) such that minX <= x <= maX and minY <= y <= maxY, right?
So if minX > maxX then the range represents an empty area of the plane (and the same with y's).
Also, as soon as `range.addPoint` has been called at least once, then `range.empty` will return false (whatever the previous state was).

How is `empty` not a good description then?

bhennion commented:
> But I have also a question, did you look at the gtk4 drawing-API? They want to replace Cairo with a node based, approach. Therefore, your changes should not increase the porting effort.

I don't think this PR influences (in any direction) the difficulty of porting to GTK4. Basically, this just regroups all the drawing operations in view/, but does not change anything to how the painting is done.

There will be more reflection to do when it'll be time to refactor XojPageView itself. I don't quite get how to replace the current buffer (a cairo_surface_t) when passing to GTK4.

Febbe commented on src/core/view/CoordinateChange.h:
```diff
@@ -0,0 +1,38 @@
+public:
+    CoordinateChange() = default;
+    CoordinateChange(const utl::Point<double> origin, double scaleFactor);
+    ~CoordinateChange() = default;
```
Remove default destructor for finalized type
```suggestion
```

Febbe commented:
For now, Cairo still exist in gtk4, it just requires an extra indirection and must be added as node into the drawing tree. But it might get deprecated some time.   
The larger problem is, that gtk4 globalized input. Now every input is caught globally and is fed into gtk-input-controlers, which then determine, if they have to accept the event. Multiple Controllers can accept the same event, and it is also incompatible with the current approach. The old approach to get events from a widget does not work anymore. Widgets don't emit events.

bhennion commented:
> For now, Cairo still exist in gtk4, it just requires an extra indirection and must be added as node into the drawing tree. But it might get deprecated some time.

As long as [GtkDrawingArea](https://docs.gtk.org/gtk4/class.DrawingArea.html) relies on cairo for its draw function, we are safe.

> The larger problem is, that gtk4 globalized input. Now every input is caught globally and is fed into gtk-input-controlers, which then determine, if they have to accept the event. Multiple Controllers can accept the same event, and it is also incompatible with the current approach. The old approach to get events from a widget does not work anymore. Widgets don't emit events.

This seems like a problem for the input system rather then for our tool handlers, does it not?

Febbe commented:
> This seems like a problem for the input system rather then for our tool handlers, does it not?

Those tool handlers rely on, or are part of, the input system.

bhennion commented:
> Those tool handlers rely on, or are part of, the input system.

Of course they rely on it. My point was: whatever is in gui/inputdevices should play the role of an abstraction layer, so that whatever relies on it should not be touched by any changes in the library.

bhennion commented:
I refactored the view classes a little bit, for better code sharing between this and the classes to come in #4158.

Technius commented:
This is rather large, so I'll take a look over the week.

bhennion commented:
Just rebased after the preliminary code cleanups were merged into master. I also added some documentation in the code.
@Technius Thanks a lot!
I think some files deserve more attention than the rest of them: namely util/DispatchPool, core/control/BaseShapeHandler and core/view/overlays/*. This is where the actual View-Controller stplitting happens.

bhennion commented:
To help with the reviewing of this PR, as well as #4158, #4159 and future PR's, here is a very coarse diagram of the intended structure:
![ViewRefactor](https://user-images.githubusercontent.com/17818041/181217282-e55362d2-168b-4d79-b087-f71aa8eb30f4.png)

This PR, and #4158 and #4159 concern only the blue (OverlayView's) and green (Overlay's, including InputHandler -- which IMO should be renamed) portions.


Now, in the last commit [e85c908](https://github.com/xournalpp/xournalpp/pull/4137/commits/e85c908690ba31d743d156fd98d11b5013d985ca), I fixed a blinking problem upon shape finalization. I do not like the way this is done, so I'd be delighted if someone had a better idea. I'll describe the issue in detail in a second post.

bhennion commented:
@xournalpp/core As I explained in #3579, I made a mistake then and never rechecked. Let it be a lesson for me.

In any case, I need to remove the mutexes from this PR. Don't review it just yet... I'm sorry about that.

bhennion commented:
> In any case, I need to remove the mutexes from this PR. Don't review it just yet... I'm sorry about that.

Well... Maybe I should not remove the mutexes just yet. There may be multiple threads involved if and when we allow several views of the same document.

Technius commented on src/core/view/overlays/BaseShapeOrSplineToolView.cpp:
```diff
@@ -0,0 +1,84 @@
+        const double* dashes = nullptr;
+        int dashCount = 0;
+        this->lineStyle.getDashes(dashes, dashCount);
+        assert((dashCount == 0 && dashes == nullptr) || (dashCount != 0 && dashes != nullptr));
```
Same concern here about the assertion.

Technius commented on src/core/view/overlays/BaseStrokeToolView.h:
```diff
@@ -0,0 +1,61 @@
+namespace xoj::view {
+class Repaintable;
+
+class BaseStrokeToolView: public OverlayView {
```
Needs a doc comment.

What are intended subclasses of `BaseStrokeToolView` other than the `BaseShapeOrSpineToolView` family?

bhennion replied:
There will be StrokeToolView, in #4158

Technius commented on src/core/view/overlays/BaseStrokeToolView.h:
```diff
@@ -0,0 +1,61 @@
+    double strokeWidth;
+
+
+    class Mask {
```
Needs a doc comment.

Technius commented on src/core/view/overlays/BaseStrokeToolView.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+BaseStrokeToolView::Mask::Mask(int width, int height, int xOffset, int yOffset, double scale) {
+    cairo_surface_t* surf = cairo_image_surface_create(CAIRO_FORMAT_A8, width, height);
```
For performance reasons, we might want to make it possible to use something like [`cairo_surface_create_similar()`](https://www.cairographics.org/manual/cairo-cairo-surface-t.html#cairo-surface-create-similar), otherwise we might be forcing the image buffer to be allocated in Xournal++'s heap instead of e.g. the display server or VRAM etc.

bhennion replied:
TBH, I have no idea which of cairo's surfaces are stored where. The documentation does not mention this at all.
They also do not mention performance differences. Do you have a source (or benchmarks, or so) that shows a difference?
In an ideal world, I agree that `cairo_surface_create_similar` should give us the best surface for the job...

Performance aside, the downside of `cairo_surface_create_similar` is that we need a cairo_surface_t instance upon creation of the mask.
I guess this can be done, by initializing the mask during the first call to ShapeToolView::draw.

bhennion replied:
I just pushed a commit making the masks through `cairo_surface_create_similar` and printing out the surface type in a terminal.
FYI, on my laptop: KDE X11 gives xlib surfaces while KDE Wayland gives image surfaces.

I'd be interested if you get different kind of surfaces depending on the OS, DE and x11 or wayland. To get this information in a terminal, you need to draw a filled highlighter shape (e.g. highlighter, with filling, rectangle).

rolandlo replied:
On Ubuntu 22.04 wayland I get `Creating mask of type: image`

rolandlo replied:
On Ubuntu 22.04 x11 I get `Creating mask of type: xlib`

rolandlo replied:
On MS Windows 10 I get `Creating mask of type: win32`

rolandlo replied:
On MacOS Catalina I get `Creating mask of type: quartz`
This may change when Gtk 3.24.35 is released and updated in `gtk-osx`.

bhennion replied:
Thanks a lot @rolandlo for the data!

I run some benchmarks I found [here](https://openbenchmarking.org/test/pts/cairo-demos) (only relevant for X11, I think), and toyed around. It turns out, the benefits of using `cairo_surface_create_similar` over `cairo_surface_create_similar_image` (on xlib surfaces) are enormous. For instance
```
$./flower-demo-xlib --benchmark
flowers: 125.03 fps
```
vs
```
$./flower-demo-image --benchmark
flowers: 14.42 fps
```
I don't quite know where things get so much slower, probably the under-use of graphics acceleration.

Anyway, this seems to be saying that we need to replace most occurences of `cairo_image_surface_create` by a `cairo_surface_create_similar`, as @Technius was suggesting. It will be pretty complicated to know the kind of surfaces we need in all the places we use image surfaces though.

I'm getting out of the PR's scope :)

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,80 @@
+    using pool_type = xoj::util::DispatchPool<T>;
+
+public:
+    void registerToPool(const std::shared_ptr<pool_type>& newpool) {
```
Doc comment should mention that this will remove it from the current pool.

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,80 @@
+};
+
+template <class T>
+class Listener {
```
Needs a doc comment.

Is my understanding of this correct: Listener is a CRTP style class. An example implementation would be `class MyListener : public Listener<MyListener> { ... }`. Subclasses of Listener are required to implement `on()` (which can potentially be overloaded), which can then be invoked by calling the `DispatchPool<MyListener>::dispatch()` method. The fun part is that because this is templated, we can abuse types like `std::enable_if_t`.

However, it seems like a concrete subclass can only implement `Listener<...>` once. Is this a problem, or do you think the overlays/views will each only be implementing one listener?

bhennion replied:
> Is my understanding of this correct: Listener is a CRTP style class. An example implementation would be `class MyListener : public Listener<MyListener> { ... }`. Subclasses of Listener are required to implement `on()` (which can potentially be overloaded), which can then be invoked by calling the `DispatchPool<MyListener>::dispatch()` method. The fun part is that because this is templated, we can abuse types like `std::enable_if_t`.

Yes, you are correct.

> However, it seems like a concrete subclass can only implement `Listener<...>` once. Is this a problem, or do you think the overlays/views will each only be implementing one listener?

If a class needs to be two kinds of listeners, you can do that
```
class A : public Listener<A> {...};  // define virtual functions void on(...); in here
class B : pulbic Listener<B> {...};  // define virtual functions void on(...); in here - types must differ from the ones in A
class C : public A, public B {...};    // overload everything here
```
This ain't a big deal, because both `A` and `B` should be also used separately elsewhere (otherwise, don't split...).

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,80 @@
+
+namespace xoj::util {
+template <class T>
+class DispatchPool final {
```
Needs a doc comment.

I initially found the type parameter name `T` to be confusing, until I saw the `using listener_type = T` line below. I would suggest renaming `T` to something more understandable like `ListenerT`.

Technius reviewed (commented):
Thanks for the work, this is sorely needed. I think the overall design looks reasonable. I haven't checked everything thoroughly, but I did look over most of the major parts.

* I have a few questions about the DispatchPool/Listener thing, as well as some of the overlay code (see below).
* I mentioned some spots that I think need doc comments, but it would probably be better to defer writing those until the overall design is finalized.

bhennion commented on src/core/view/overlays/BaseShapeOrSplineToolView.cpp:
```diff
@@ -0,0 +1,84 @@
+        const double* dashes = nullptr;
+        int dashCount = 0;
+        this->lineStyle.getDashes(dashes, dashCount);
+        assert((dashCount == 0 && dashes == nullptr) || (dashCount != 0 && dashes != nullptr));
```
Yes to everything!

bhennion commented on src/core/control/tools/RectangleHandler.cpp:
```diff
@@ -4,72 +4,66 @@
+    Range rg(p1.x, p1.y);
+    rg.addPoint(p2.x, p2.y);
+
+    return {shape, rg};
```
I was counting on some (hypothetical) copy elision kind of compiler optimisation here. I assumed that an `std::move` would block such optimisation.
I can always make the pair before, and emplace points into `first`, but that's definitely less readable.

bhennion commented on src/core/control/tools/EllipseHandler.cpp:
```diff
@@ -2,95 +2,84 @@
+
+    // set resolution proportional to radius
+    auto npts = static_cast<size_t>(std::round((std::abs(radiusX) + std::abs(radiusY)) * 2.0));
+    if (npts < 24) {
+        npts = 24;  // min. number of points
+    }
```
Changed by an optimal formula for the number of points.

bhennion commented:
Here are the last changes I made:

1. Split off the mask class to a dedicated file. Then, in future PR's, we can use it for every mask or surface buffer (e.g. in StrokeView, PageView, VerticalToolHandler).
2. I removed the mutex. As I mentionned above, there is actually oly one thread involved. The structure is not changed, so putting the mutex back in if we need to later won't be hard at all.

This is ready for another round of review, for whoever feels like it :)

Technius commented on src/core/view/StrokeViewHelper.cpp:
```diff
@@ -0,0 +1,64 @@
+    assert((dashCount == 0 && dashes == nullptr) || (dashCount != 0 && dashes != nullptr));
+
+    if (dashes) {
+        for (const auto& [p, q]: PairView(pts)) {
```
Makes sense. Although for the branching issue, the pattern seems simple enough that modern CPU branch predictors should be able to optimize it.

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,111 @@
+namespace xoj::util {
+
+/**
+ * @brief Generic template for dispatchers.
```
```suggestion
 * @brief Utility class for dispatching a method call to multiple "listeners".
```

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,111 @@
+public:
+    using listener_type = ListenerT;
+
+    template <typename... Args>
```
Needs a doc comment.
```suggestion
    /**
     * @brief Invokes the `on()` method of all registered `ListenerT`s.
     */
    template <typename... Args>
```

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    void add(listener_type* v) {
```
```suggestion
    /**
     * The listener cannot already be registered.
     */
    void add(listener_type* v) {
```

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -0,0 +1,111 @@
+ *      void on(...);  // Signal receiver
+ *  };
+ *
+ * WARNING: Always unregisterFromPool() in derived class destructor.
```
This seems like a pretty big caveat. Would it be a big problem to add a virtual destructor in Listener that automatically does this?

bhennion replied:
I'd rather not impose a vtable on every listener type.
More importantly, this would open a data race upon destruction:
the base class gets destroyed last, so in a multithreading situation, a function on() of the derived class could be called (by the dispatcher) between the destruction of the derived class and before the destruction of the base (hence the call to unregisterFromPool()).

I don't have a proper solution for this.

Technius commented on src/core/view/Mask.cpp:
```diff
@@ -0,0 +1,148 @@
+
+using namespace xoj::view;
+
+// #define DEBUG_MASKS
```
```suggestion
```

Technius commented on src/core/view/Mask.h:
```diff
@@ -0,0 +1,34 @@
+
+namespace xoj::view {
+
+class Mask {
```
Needs doc comments.

Thanks for adding this, we could probably simplify several of the tool handlers with this.

In a follow-up: it would also be useful to have the ability to resize the cairo_surface (such as is done in VerticalToolHandler).

bhennion replied:
> In a follow-up: it would also be useful to have the ability to resize the cairo_surface (such as is done in VerticalToolHandler).

Yes I agree. Also in the other tools handlers actually: the tool handlers should listen to zoom or scroll events and adapt their masks accordingly, like the VerticalToolHandler does.

bhennion replied:
Also: This class could be placed in util rather than core/view. Not sure which makes more sense.

rolandlo commented on src/core/model/Stroke.cpp:
```diff
@@ -302,6 +302,22 @@ Point Stroke::getPoint(PathParameter parameter) const {
+        // We cannot deduce the bounding box from the snapping box if the stroke has pressure values
+        this->sizeCalculated = false;
+    } else {
+        assert(snappingBox->isValid());
```
This assertion is not satisfied for a horizontal or vertical straight line (drawn with the "Line" drawing type, with grid snapping on) and the app crashes.

bhennion replied:
Thanks for the catch! I'll get back to this PR in a moment.

bhennion commented:
Hey there!

I cleaned up the includes here and there, and tidied the commit tree. @Technius @Febbe @LittleHuba: would any of you give this another review or do you think/trust it is good to go?

Technius commented on src/core/gui/PageView.h:
```diff
@@ -61,12 +61,19 @@ class XojPageView: public LegacyRedrawable, public PageListener, public xoj::vie
 
     double getWidth() const override;
     double getHeight() const override;
+    //
```
nit: unnecessary comment

Technius commented on src/core/view/Repaintable.h:
```diff
@@ -41,5 +42,11 @@ class Repaintable {
     virtual void flagDirtyRegion(const Range& rg) const = 0;
+
+    /**
+     * @brief Remove a tool view repaint the given range
```
```suggestion
     * @brief Remove a tool view and repaint the given range
```

Technius commented on src/core/view/overlays/ShapeToolView.h:
```diff
@@ -43,6 +43,14 @@ class ShapeToolView final: public BaseShapeOrSplineToolView, public xoj::util::L
+    /**
+     * @brief Called before the corresponding BaseShapeHandler's destruction
+     * @param rg The bounding box of the entire stroke
+     */
```
Shouldn't the doc comment go on the function? Also, I think the `Finalization` struct itself also needs a doc comment.

Technius commented on src/util/include/util/DispatchPool.h:
```diff
@@ -39,6 +39,21 @@ class DispatchPool final {
+        // leading to possible segfaults.
+        auto p = std::move(this->pool);
+        for (auto* v: p) {
+            v->deleteOn(std::forward<Args>(args)...);
```
Unless I'm missing something, shouldn't `deleteOn` also be on the list of required methods in the class's doc comment?

Technius reviewed (approved):
I assume that you didn't modify anything when you rebased and only looked at the last too commits. LGTM if so, besides one small assert issue + some minor documentation issues.

bhennion commented:
> I assume that you didn't modify anything when you rebased

That is correct!

I pushed a fixup commit with your comments. Thanks for the review!!

Should we plan on merging this then?

Technius commented:
Yes, I think this can be merged unless other people have any objections.

bhennion commented:
Merging this in 5 days, unless someone objects.
